osquery/tables/ipmi/ipmi_info.cpp

```diff
@@ -0,0 +1,1297 @@
+  if (rv != 0) {
+    LOG(WARNING) << "Error setting up SMI connection: " << strerror(rv);
+    return;
+  }
+
+  TLOG << "Opening IPMI domain";
+  if (ipmi_open_domain(
+          "", &con_, 1, NULL, NULL, IPMIFullyUpCB, this, NULL, 0, NULL) != 0) {
+    LOG(ERROR) << "Error opening IPMI domain: " << strerror(rv);
+    return;
+  }
+
+  TLOG << "Adding IPMIClient as a dispatcher service";
+  Dispatcher::addService(
+      std::shared_ptr<IPMIClient>(this, [](IPMIClient* c) { c->stop(); }));
+
+  // Set default timeout value.
+  size_t timeout = 360000;
+  auto parser = Config::getParser("ipmi");
+  if (parser != nullptr && parser.get() != nullptr) {
+    const auto& config = parser->getData().get_child(
+        kIPMIConfigParserRootKey, boost::property_tree::ptree("UNEXPECTED"));
+    if (config.get_value("") == "UNEXPECTED") {
+      LOG(WARNING) << "Could not load ipmi configuration root key: "
+                   << kIPMIConfigParserRootKey;
+    }
+
+    timeout = config.get_child("timeout", boost::property_tree::ptree())
+                  .get_value<size_t>(360000);
+  }
+
+  TLOG << "Waiting for OpenIPMI to reach 'fully up' state to retrieve "
+          "domain (may take up to "
+       << timeout << " ms)";
+  blkAndOp([this] { return domain_ != nullptr; }, timeout);
+
+  if (domain_ != nullptr) {
+    up_.store(true);
+  } else {
+    LOG(WARNING) << "IPMI BMC was not responsive";
+    return;
+  }
+
+  TLOG << "Searching for IPMI channel for LAN communication";
+  findLANCh();
+
+  TLOG << "Setting up ipmi_lanparm_t for system MCs";
+  rv = ipmi_domain_iterate_mcs(
+      domain_,
+      [](ipmi_domain_t* domain, ipmi_mc_t* mc, void* data) {
+        IPMIClient* c = reinterpret_cast<IPMIClient*>(data);
+        c->addLANParm(getMCKey(mc), mc);
+      },
+      this);
+  if (rv != 0) {
+    LOG(ERROR) << "Could not register ipmi_domain_iterate_mcs: "
+               << strerror(rv);
+  }
+
+  blkAndOp([]() { return false; }, 1000);
+
+  TLOG << "IPMIClient initialization complete";
+}
+
+ipmi_lanparm_t* IPMIClient::getLANParm(const std::string& name, ipmi_mc_t* mc) {
+  if (lanparms_.find(name) != lanparms_.end()) {
+    return lanparms_[name].get();
+  }
+
+  if (addLANParm(name, mc)) {
+    return lanparms_[name].get();
+  }
+
+  return nullptr;
+}
+
+bool IPMIClient::addLANParm(const std::string& name, ipmi_mc_t* mc) {
+  ipmi_lanparm_t* lp = nullptr;
+
+  int rv = ipmi_lanparm_alloc(mc, lanCh_, &lp);
+  if (rv != 0) {
+    LOG(ERROR) << "Could not allocate ipmi_lanparm_alloc for local BMC: "
+               << strerror(rv);
+    return false;
+  }
+
+  if (lp == nullptr) {
+    LOG(ERROR) << "Could not successfully allocation "
+                  "ipmi_lanparm_t*: still a nullptr";
+    return false;
+  }
+
+  lanparms_[getMCKey(mc)] =
+      std::unique_ptr<ipmi_lanparm_t, std::function<void(ipmi_lanparm_t*)>>(
+          lp, kFreeLANParm);
+  return true;
+}
+
+void IPMIClient::insertRowsQueue(std::string key,
```

Comment:
This might've already been mentioned above, but can this be `const std::string& key` as you don't seem to be modifying the value?